### ISSUE TITLE
job-manager: add priority plugin interface

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#3311]
A secondary priority value for jobs was described in #3256.

Presumably we would develop a job manager plugin of some sort to generate updates to this secondary priority based on multiple factors, one of which is the per-user fair share value from flux-accounting.

This issue is about specifying a mechanism for flux-accounting to communicate periodic updates of per-user fair share values to flux-core.

**Comment by garlick** [comment]
Here is one straw man proposal after thinking a bit about our conversation this afternoon.

Have the job manager keep a "active users" hash, with use counts per active job, and a key-value hash per user.  RPCs could be implemented to:
* list the current active users
* set key=value for a given user
* perform a bulk update of key=value for multiple users

The schema of keys could be unknown to the job manager, but a priority plugin, such as one that is fair share aware, could read specific per-user keys and involve them in calculation of a priority update.

When the job manager unloads, it could dump this user hash to the KVS, and reload it when it starts up again.

When flux accounting wants to recalculate its fair share values (say every hour via a cron job) it could
* ask job-manager for current active user list
* perform bulk update of fairshare=value for those users (future optimization: only those who have changed)

After the update, the plugin could run to recalculate the secondary priority for jobs in the queue.

**Comment by garlick** [comment]
That was kind of an overly detailed, job-manager centric thought bubble.  Sorry about that.

If I could boil it down, I think my main thought above was that the job manager could implement a generic mechanism, such as a key/value store, for caching user data needed by a priority plugin, and that an external program (e.g. run by cron) could push that data into the job manager, limiting its updates to only the current active users.  So the external program and an associated priority plugin would be aware of a particular "schema" for the data and control its update frequency.  The job manager would be unaware of that schema.  It would simply trigger the configured plugin to run when any of its inputs might have changed (possibly handing it a list of users and/or jobs).  The plugin's updates to the secondary priority would in turn possibly trigger a move of a job within the queue.

I might still be too bogged down in the details (sorry!)

**Comment by grondo** [comment]
An alternative would be to cache data within the plugin itself. job-manager plugins could have a `init` callback which would (similar to shell plugins) allow them to register a service endpoint under the job-manager, only if necessary. Then the external part of the fairshare priority mechanism would send updates directly to this handler. This would allow job-manager plugins to keep state in a manner of their choosing. Other priority plugin types could setup other mechanisms for updating priority, e.g. set a timer_watcher, etc. Plugins could register a blob for the job-manager to store in the KVS when it is unloaded, or they could be required to handle this themselves.

The benefit here is less touching of the job-manager code, and better abstraction for priority plugins. The plugin architecture could even allow multiple plugins to be loaded at once for generically extending job-manager functionality. (i.e. priority plugin is just one type of job-manager plugin)

This approach may also allow more rapid development of the initial multifactor priority plugin since we do not first need to develop a generalized data caching mechanism for the job-manager proper.

**Comment by garlick** [comment]
That makes sense to me.

A requirements question is: do jobs need to have accurate secondary priorities assigned before they become eligible for scheduling?

**Comment by grondo** [comment]
Good question for @dongahn, but I would assume so. Then perhaps callback for priority plugin should be made before jobs are inserted into the queue to calculate initial secondary priority using current info.

**Comment by dongahn** [comment]
I think so. If the instance is configured to use the secondary priorities, a job must be assigned to a value before being scheduled.

**Comment by dongahn** [comment]
> perform bulk update of fairshare=value for those users (future optimization: only those who have changed)

@grondo, @SteVwonder, @cmoussa1 and @milroy discussed this a bit at today's coffee hour.

Currently the fair share value is essentially equal to the global rank of users. So it is likely the fairshares of a majority of users will change for common cases at every invocation of fshare calculation. E.g., if the previously highest-priority job becomes the lowest priority for the current round, this single change will change the rank of every user.

Once we make reasonable progress with unoptimized update protocol, there are some techniques we can look into. I briefly mentioned this yesterday but perhaps the concept of the edit distance can be expanded. Introducing floating point schemes such a way that we minimize the previously calculated fairshare values can also be fruitful. As is, We're tied to the normalized fair share values between [1.0, 0), which won't work well to support such an augmented scheme.

In flux-accounting, I will open up a ticket to augment `libweighted_tree` to incur minimal fairshare value changes.

BTW, @SteVwonder correctly noticed that solving this may also allow us to incur minimal changes on subsequent job-manager to fluxion-qmanager updates.

**Comment by dongahn** [comment]
@garlick and @grondo: For unoptimized update protocol, once the weighted tree walks the tree, it will have a sorted vector of users:

For example, in this simple test code

walk: https://github.com/flux-framework/flux-accounting/blob/master/src/fairness/weighted_tree/test/weighted_tree_test01.cpp#L78

iteration of ordered user vector: https://github.com/flux-framework/flux-accounting/blob/master/src/fairness/weighted_tree/test/weighted_tree_test01.cpp#L84

Perhaps we can agree on the proper format of this order user vector (that goes into flux-core) so that a program can be developed at flux-accounting in the near future? For example, essentially the payload schema for the upcoming RPC or similar? 

I guess we may need to decide whether plugin will do this or job-manager will do this to decide on the schema... Or would it matter when the schema is likely a key-value set?

**Comment by grondo** [comment]
> I guess we may need to decide whether plugin will do this or job-manager will do this to decide on the schema... Or would it matter when the schema is likely a key-value set?

Yes, I think this is the first design decision we should make. My opinion would be that we design an interface for a generic plugin architecture for the job-manager (perhaps priority-specific at this point) with an interface for plugins to register callbacks in the job-managers flux_t handle. This would allow the plugin to independently be developed in parallel (and wholly in flux-accounting) with the job-manager plugin interface in flux-core.

In any event, it would seem you could make considerable progress on priority plugin building blocks which could be tested using unit tests. Maybe seeing this development would inform a better design of the priority plugin interface in the job-manager. I still feel a bit in the dark on how the whole process fits together (e.g. job completion data from job-archive goes into accounting database, something calculates global fairshare vector, somehow the output of this vector creates a factor that is used in a multi-factor calculation to augment existing job priority values). It would allow flux-accounting to make good progress if you had some way to drive this process with mock data in flux-account unit and/or system tests.

**Comment by dongahn** [comment]
> In any event, it would seem you could make considerable progress on priority plugin building blocks which could be tested using unit tests. 

https://github.com/flux-framework/flux-accounting/pull/65 is the big part of this building blocks including unit tests. When you have a chance,  please take a look at the unit tests (https://github.com/flux-framework/flux-accounting/pull/65/commits/746a6ce306ef164907b0c883c8031436ee63fd15) to let us know if it is along the line of  what you are thinking about. 

How about for me or @cmoussa1 to take the next step to create a mock program that uses this library and outputs a JSON with order vector of users? The program can either pass the output as an output file or similar (how do you do this with python job validator?) or sends the vector through an yet-to-be developed RPC -- in this case it would use `flux_open`


>  I still feel a bit in the dark on how the whole process fits together (e.g. job completion data from job-archive goes into accounting database, 

This is the block that  @cmoussa1 is developing under the guidance of @chu11. He should chime in with where he is.

> something calculates global fairshare vector, somehow the output of this vector creates a factor 

This is the block (`libweighted_tree`) that was developed as above. 

> that is used in a multi-factor calculation to augment existing job priority values).

This is the block that should be (co-)architected into flux-core, I think.

**Comment by grondo** [comment]
This sounds good to me @dongahn! I will take a look at https://github.com/flux-framework/flux-accounting/pull/65. Thanks!

**Comment by dongahn** [comment]
>> flux-framework/flux-accounting#65 is the big part of this building blocks including unit tests. When you have a chance, please take a look at the unit tests (flux-framework/flux-accounting@746a6ce) to let us know if it is along the line of what you are thinking about.

> How about for me or @cmoussa1 to take the next step to create a mock program that uses this library and outputs a JSON with order vector of users? The program can either pass the output as an output file or similar (how do you do this with python job validator?) or sends the vector through an yet-to-be developed RPC -- in this case it would use flux_open

@cmoussa1: I can make today's 2PM coffee hour for the first 30 mins. Perhaps we should discuss the initial step there? Now that all of the PRs have been landed for flux-accounting, it would be good to generate what need to be done for the next release cycle.

**Comment by cmoussa1** [comment]
> > I still feel a bit in the dark on how the whole process fits together (e.g. job completion data from job-archive goes into accounting database,

> This is the block that @cmoussa1 is developing under the guidance of @chu11. He should chime in with where he is.

Before we shifted to working on first implementations for the weighted tree library, I was working on implementing a job usage calculation that utilizes @chu11's job-archive module (and that eventually makes its way into the flux-accounting database). I think I've made some good progress there, but I am sure there will be good feedback and suggestions to further optimize it. Maybe this week during the flux-accounting meeting I can talk about where I am at and see if this is indeed the right course of action. 

> @cmoussa1: I can make today's 2PM coffee hour for the first 30 mins. Perhaps we should discuss the initial step there? Now that all of the PRs have been landed for flux-accounting, it would be good to generate what need to be done for the next release cycle.

@dongahn - that sounds like a good plan. Talk to you then.

**Comment by garlick** [comment]
Let me try to summarize the architecture we discussed in today's 2pm coffee call, or how I am picturing it.  Please comment!

The job manager would offer a "priority plugin API".   This would be a published API, similar to the job shell plugin API, that would enable out of tree projects to provide job manager plugins that (a) can access various job attributes, and (b) can use that info, and other info perhaps, to set/update, the secondary priority of individual jobs.

The secondary priority would, in turn, be used by the job manager and schedulers to order the queue of alloc requests.  Specifically, the order would be determined by (in descending precedence): 1) administrative priority, 2) secondary priority, 3) submit time.  _Aside: actually let's revisit this.  Perhaps there should only be one priority that is calculated by the plugin and takes 1) and 3) as input._

The API will need to include a mechanism for the job manager to ask the plugin for an initial priority priority value for a new job.  I think this "request" from the job manager to the plugin would need to occur and be satisfied on transition of a job on entry to SCHED state, since the priority is required to be set before an alloc request is sent to the scheduler.

There would be another API interface that would allow the plugin to asynchronously update the secondary priority of a pending job, much like `flux job priority` allows the administrative priority to be updated.

Other plugin API interfaces that I think we can project will be necessary:
* get job attribute: e.g. submit time, admin priority, job owner, current secondary priority, jobspec (resources section)
* notification of jobs that are no longer pending
* a mechanism for registering RPC sevices like the job shell, e.g. `job-manager.priority-<method>`.

We discussed that a priority plugin that uses the above interface to implement a "fair share" specific multi-factor priority calculation could be part of the flux-accounting project.  Its priority calculation would take as inputs some of the above data from the job manager plugin API, and some data from flux-accounting, especially the user's fair share factor.  This info would be pre-loaded by the plugin on initialization.  Furthermore, since the fair share factors evolve over time, the plugin would need to set up a service method to accept updates periodically.

It would be useful to further flesh out the data that a priority plugin might require.  We have a start on the factors needed for a fair share based multi-factor calculation in flux-framework/flux-accounting#8 although it is pretty slurm specific and needs to be fluxified.

**Comment by dongahn** [comment]
Thanks @garlick. This sounds like a great start.

> The API will need to include a mechanism for the job manager to ask the plugin for an initial priority priority value for a new job. I think this "request" from the job manager to the plugin would need to occur and be satisfied on transition of a job to SCHED state, since the priority is required to be set before an alloc request is sent to the scheduler.

Perhaps we will need support for no plug-in or dummy plug-in support. For nested instances where plug-ins are not provided, job manager still want to transition a job to SCHED with no or equal secondary priority.
 
> get job attribute: e.g. submit time, admin priority, job owner, current secondary priority, jobspec (resources section)

It seems we need some thoughts about those job attributes that job-manager may not know immediately. Currently we set things like `queue` in the attr section of the job spec. Maybe the job spec can be passed to the multi factor priorities to help it harvest those scheduler specific attributes as well. 

> There would be another API interface that would allow the plugin to asynchronously update the secondary priority of a pending job, much like flux job priority allows the administrative priority to be updated.

Does it mean the multi-factor priority plugin needs to keep track of all of pending jobs? I initially thought that we can avoid this, by designing this asyn interface to operate at the user account level. But then job-manager probably doesn't even want to have the concept of bank accounts/users, so perhaps replicating job list within this plugin is unavoidable...

**Comment by garlick** [comment]
Good comments!

> For nested instances where plug-ins are not provided, job manager still wants to transition a job to SCHED with no or equal secondary priority.

Agreed.  I was assuming that the plugin would not be loaded by default and things would work as they do now.  The system instance could be explicitly configured to load the fair share plugin.  Alternatively (see _aside_ above), we could load a default plugin that calculates a single priority value, taking as input the administrative priority and submit time.

> It seems we need some thoughts about those job attributes that job-manager may not know immediately. Currently we set things like queue in the attr section of the job spec. Maybe the job spec can be passed to the multi factor priorities to help it harvest those scheduler specific attributes as well.

Passing the jobspec through may be a good option.  I was thinking we would just pass the resources section to avoid the long environment but attributes surely would be needed too.  Note that the job manager currently doesn't have the jobpsec in hand for each job so that may need some work.

> Does it mean the multi-factor priority plugin needs to keep track of all of pending jobs?

Hmm, maybe, or maybe we could implement a query at the plugin API level for listing all jobs by owner?

**Comment by dongahn** [comment]
> Hmm, maybe, or maybe we could implement a query at the plugin API level for listing all jobs by owner?

I like this idea. But in this case, we may need some good performance requirement for the query to return the list jobs own by the user. BTW, is the assumption here the user account seen by flux-accounting will be the same user ID maintained by flux-core?

**Comment by garlick** [comment]
> we may need some good performance requirement for the query to return the list jobs own by the user.

Excellent point.  If this plugin runs in the job manager's thread (what I was proposing) then it should not hold onto it for long or it will negatively impact job throughput.  It will get control through its update RPC, when the job manager calls in to get priority for a new job, or when any of its own reactor watchers run.   Any API call we offer should be designed to be fast even when the number of pending jobs is large.

When the plugin has a significant batch of work such as processing a new vector of fair share values for multiple users, it may need to interleave its work with letting the job manager run, for example using the prep/check/idle reactor idiom.

> BTW, is the assumption here the user account seen by flux-accounting will be the same user ID maintained by flux-core?

Yes I think that is OK for now.  At some point when the flux-accounting db is multi-cluster, and multiple sites are using it, we may need to map to an organization's unique ID in some way, but I don't think we need to worry about that right now.

**Comment by dongahn** [comment]
> a mechanism for registering RPC sevices like the job shell, e.g. job-manager.priority-<method>.

Is this so that the multi-factor priority (MFP) plugin client can do an RPC to the MFP plugin server or something else?

**Comment by dongahn** [comment]
> It would be useful to further flesh out the data that a priority plugin might require. We have a start on the factors needed for a fair share based multi-factor calculation in flux-framework/flux-accounting#8 although it is pretty slurm specific and needs to be fluxified.

I'd ask @cmoussa1 to propose a minimal set that we should tackle for the first round. As part of that, we should get the terms right to be more appropriate for Flux.

**Comment by grondo** [comment]
> The API will need to include a mechanism for the job manager to ask the plugin for an initial priority priority value for a new job. I think this "request" from the job manager to the plugin would need to occur and be satisfied on transition of a job to SCHED state, since the priority is required to be set before an alloc request is sent to the scheduler.

At first I was wondering if there was really a need for a separate call for the initial priority. If priority is a function of job parameters + some internal plugin state, then a single get-priority call should work the same for the initial priority vs updates.

However, on second thought, it might make things easier for a plugin developer to have a callback as a job-manager job enters each main job state (i.e. "init", "depend", "sched", "run", "cleanup", "inactive"). This would allow a priority plugin to hook in to the correct place to initialize its internal state (e.g. sometimes a priority plugin would want to create internal state for jobs even if they are in DEPEND, or keep that state while they are in CLEANUP). Any plugin callback after SCHED would not be able to update job data such as the priority, but a sophisticated plugin could use this data to update internal state that influences the priority of pending jobs). At all times, the "inactive" callback could be used to delete any internal job state (such as a count of running jobs for a given user, etc)

This kind of implies that the job-manager would drive the recalculation of priorities on a per-job basis, e.g. the job priority plugin interface would have callbacks that pass in a single job. Perhaps when internal state in a priority plugin has been updated, the plugin can set a flag in the job-manager to request a re-prioritization loop, which would be driven at the job-manager's discretion? 

Sorry this may have gotten a little off-topic.

**Comment by grondo** [comment]
> > a mechanism for registering RPC sevices like the job shell, e.g. job-manager.priority-.
>
> Is this so that the multi-factor priority (MFP) plugin client can do an RPC to the MFP plugin server or something else?

Yes, AIUI, the idea here is to allow the plugin to accept an update of data from an external source. In this case a provided script or cron job could periodically push updates of one or more factors to the plugin. The service registered by the plugin could also respond to a request for the current data values for debugging or informational purposes.

**Comment by garlick** [comment]
> This kind of implies that the job-manager would drive the recalculation of priorities on a per-job basis, e.g. the job priority plugin interface would have callbacks that pass in a single job. Perhaps when internal state in a priority plugin has been updated, the plugin can set a flag in the job-manager to request a re-prioritization loop, which would be driven at the job-manager's discretion?

Ah I was thinking that the plugin would be required to select which jobs need to be updated.  If we wanted to just let the plugin trigger a job manager iteration over all pending jobs when it receives updated factors, that would simplify things quite a bit:
* no fast `getjobsbyuser()` lookup required
* responsiveness of the job manager reactor during a bulk update could be managed in the job manager using prep/check/idle or similar, not in the plugin
* a really dumb plugin could be implement the single `getpriority()` callback.

I hadn't thought of doing it that way but now that you suggest it, I really like it especially for a first cut since it will significantly simplify the plugin and its API.

On question:  Is it reasonable to require that a `getpriority()` callback must always return a result immediately, or must we design some sort of async interface that allows the plugin to make RPCs?  It would be a lot simpler design if we could assume an immediate response, but that would presume that any externally-sourced data is preloaded in the plugin at initialization (for example the fair share factors for all possible users).

**Comment by garlick** [comment]
Another way to go might be to add a new PRIORITY job state between DEPEND and SCHED.   The job would transition from PRIORITY to SCHED once its priority is established.  The priority plugin could watch for a job state transition into PRIORITY (or earlier) which would trigger it to fetch a user's data, if not in cache, and call a plugin API function to set the job's priority.

After that we could assume that a user's data is cached and drive priority updates from the job manager, expecting a `getpriority()` function to return immediately.

I guess the PRIORITY state is not really necessary and we could have another notification in SCHED state that the job needs its priority set, and hold back the alloc request until its done.  However, I kind of like the idea of a new state here as if flux-accounting turns into a center resource, it might occasionally be slow or unavailable, and having jobs stuck in a PRIORITY state would make it pretty obvious what is going on.  That also might be useful when the priority plugin is provided by a site, and the "why is flux slow" question could be answered if the time spent in the priority state is shown to be long.

**Comment by grondo** [comment]
I like the idea of the PRIORITY state. However, I think we should discourage a design where a priority plugin fetches something from a remote service or does a blocking operation in the common case to calculate the initial PRIORITY.  The plugin should have enough information from the previous out-of-band update of internal state to calculate an initial priority for any new job.

However, I do like that the PRIORITY state allows more flexibility and would allow a plugin to be developed in this way if desired!

I also agree that priority updates should be instant (no RPCs, no blocking) in a well-behaved plugin. RPCs to update internal state should be handled either out-of-band, or I suppose could be scheduled via a timer watcher (with no blocking).

**Comment by grondo** [comment]
One reason to still allow a priority plugin to have callbacks even in non-PRIORITY states is that this will allow a plugin to update some of its internal state in-band, e.g. total count of queued jobs for a user, perhaps the number of _running_  jobs for a given user, or some implementations may want to count jobs in CLEANUP, some may not, etc. By having a place to hook into the job-manager in all of these cases, we allow the plugin developer maximum flexibility in the implementation.

**Comment by garlick** [comment]
Oh yeah I meant to state agreement with the idea that it should be possible through the plugin API to register a callback on all job state transitions.  In fact  maybe this could just be a subscription to the already-defined job state transition event.

> However, I do like that the PRIORITY state allows more flexibility and would allow a plugin to be developed in this way if desired!

It would let the fair share priority plugin be implemented different ways:
* initially load fair share factors for all center users, and call `setpriority()` for a job immediately upon its entry to PRIORITY state.  In the rare case that a user is not known at the time of startup, job could remain in PRIORITY state until the next periodic update of fair share factors brings in the user
* initially load nothing, and leave jobs in PRIORITY until first periodic update of fair share factors primes the cache
* initially load nothing, and upon entry into PRIORITY, fetch the user's fair share factor from flux-accounting, calling `setpriority()` once response is received.

In the latter case, fair share factors could be timed out of the cache if desired.

**Comment by grondo** [comment]
>  In fact maybe this could just be a subscription to the already-defined job state transition event.

I would consider not using an event subscription for this. It could make plugin development less straightforward since it would introduce races, and would require the plugin to parse the job state event payload.  You would also need to provide some way for the plugin to fetch the job by id from the job-manager, whereas with a more typical plugin callback method, the job-manager can pass the job directly in to the plugin's callback method.

**Comment by grondo** [comment]
BTW, what we're moving towards here seems like it could be more powerful than just a job-manager priority plugin. If plugins have a hook into every job state transition, an interface for querying job properties including portions of jobspec, and an initialization hook that allows plugins to register message handlers and watchers on the job-manager's flux handle, then plugins could be written that do a lot more than manage administrative job priorities (just as an example, the job-manager "journal" could have likely been developed as a plugin)

As we design the interface, we could keep this in mind, though it is only a secondary use case. But if we can avoid creating priority-specific interfaces, then we keep the doors open to other job-manager extensions in the future.

**Comment by garlick** [comment]
> As we design the interface, we could keep this in mind, though it is only a secondary use case. But if we can avoid creating priority-specific interfaces, then we keep the doors open to other job-manager extensions in the future.

Sure

> (just as an example, the job-manager "journal" could have likely been developed as a plugin)

I was just about to disagree with you and point out what new hooks would be required to do that - then I realized we wouldn't even need the journal if we had done job-info as a job manager plugin.  Maybe I should have said that - @chu11 might kill me.

**Comment by dongahn** [comment]
Great discussions. In terms of making lock-step progress w/ flux-accounting, do we think it is a good idea to do a skeleton RFC for the MFP plugin API? Once we can narrow down on the design details a bit using the RFC skeleton, perhaps flux-accounting can make progress based on that without actually having this API implemented within flux-core. Then, when flux-core team can make progress on this API, we can start our next-level round co-design/-implementation?

**Comment by garlick** [comment]
My plan is to open two RFC PRs for a start:
* RFC 27 Sched Alloc Proto:  add protocol for reprioritizing jobs
* RFC 21 Job States & Events:  add PRIORITY state

We can continue discussion on those details in the PRs.

**Comment by dongahn** [comment]
@garlick: yeah this is great! Thank you for doing this for us!

**Comment by SteVwonder** [comment]
With all of our other state transitions, they implicitly represent a "hand-off" of the job from one module "owner" to another.  If we are adding a `PRIORITY` state, does it make sense to add a priority module, and then have flux-accounting provide a plugin to that module?  It's definitely not necessary, but I just wanted to mention the pattern that already exists and ask if we want to continue it.

**Comment by garlick** [comment]
That's an interesting comment.  Had to think a bit, but I believe we still are preserving the pattern for the most part.  During the PRIORITY state we are waiting for an external entity (in this case a plugin possibly from another project) which might or might not need to phone home to flux-accounting or similar, so it feels appropriate to me for that to be event driven, and for the state to be visible since it will provide some insight if the phone home part is slow or stuck.  

We could perhaps arrange things so that the plugin/service is another module that communicates only with messages, or the other option I was thinking about earlier was a python filter like the ingest validator.  But it seems like a direct plugin might be easier to reason about and more efficient, otherwise a lot of job data might need to be transferred/cached in the other module/coprocess.  Simple function calls turn into RPCs that have to be managed asynchronously, etc..

At least that's the way @grondo and I were leaning.  The benefit of going the other way would be some isolation from bugs/slowness, especially in the coprocess case which can't crash the broker even if it segfaults.

**Comment by SteVwonder** [comment]
> But it seems like a direct plugin might be easier to reason about and more efficient, otherwise a lot of job data might need to be transferred/cached in the other module/coprocess. Simple function calls turn into RPCs that have to be managed asynchronously, etc..

Yeah, I agree this is a massive overhead (developer and message-wise) to make it a separate module.  I think the reasoning is sound for making it a plugin of the job manager, just wanted to toss it out there to make sure we weren't missing anything.

**Comment by chu11** [comment]
a subtlety in the above discussion that perhaps I missed (or perhaps hasn't been determined).  When a job enters the PRIORITY state, it asks the plugin for the initial queue priority.  But what about any priority changes after that?

- Does the job-manager call the priority plugin every once in awhile? perhaps at configured intervals
- When it calls the priority plugin, does it send 1 job at a time, or all the jobs currently in the queue?  I have the feeling it has to be the latter?

**Comment by grondo** [comment]
> But what about any priority changes after that?

AIUI, we were considering two interfaces here:

 - a simple callback into the plugin which given a job returns a priority. This would be used in place of any default `job_update_priority()` or similar function in the job manager. In this case the function would be called once per job, whenever the job manager wanted to update a job's priority. A limitation would be that the function to calculate the priority is understood not to block.

 - a `setpriority()` call that the plugin could call asynchronously on any job. The job manager would then perhaps notify the priority plugin of job/jobs that need a priority update, and the plugin could either synchronously or asynchronously update the job or job's prioirty.

I would argue for the first case above initially. This approach is much simpler and allows the job-manager full control over when job priorities are updated. The plugin could update its internal state asynchronously (via RPCs, timer watchers, or a service method to accept updates), so that the `job_update_priority()` call is always just a simple calculation. The job manager plugin API could offer a method to allow the plugin to trigger a reprioritization.

For the second case, the plugin would have control over when job priorities are updated, and thus the job-manager would have to have some slightly sophisticated code to batch priority updates before acting on them, and if it wanted to trigger a priority recalculation for one or more jobs, it would need code to be able to handle this asynchronously.

**Comment by chu11** [comment]
> The job manager plugin API could offer a method to allow the plugin to trigger a reprioritization.

Sounds good.  This was the glue piece that I did not see.

I was thinking of a basic priority plugin that mirrors what the job-manager does right now, where jobs are ordered by admin priority first, `t_submit` second.  If the user changes the admin priority for a job via `flux job priority`, it could cause many jobs to have their priorities changed to ensure ordering is correct.

**Comment by grondo** [comment]
> If the user changes the admin priority for a job via flux job priority, it could cause many jobs to have their priorities changed to ensure ordering is correct.

Doesn't the job-manager use `zlistx_t` for the ordered queue? If a job has its priority updated, it should be moved simply with `zlistx_reorder()` (perhaps that is already done in the job-manager since updating priority is already supported?

**Comment by grondo** [comment]
Before any work starts on this, we should probably have proposed API for job-manager plugins and make sure it will cover our basic use cases. We can then consider how and when the job-manager will call each defined plugin callback to make sure the plugins have the required hooks, and the priority updates will work for the intended cases.

**Comment by chu11** [comment]
> Doesn't the job-manager use zlistx_t for the ordered queue?

Yes

> If a job has its priority updated, it should be moved simply with zlistx_reorder() (perhaps that is already done in the job-manager since updating priority is already supported?)

My understanding is that with the priority plugin, queue priorities become the only data used for ordering, `t_submit` is no longer used in ordering.  Possibly bad example:  a bunch of jobs are assigned queue priorities of 100, 99, 98, 97, 96.  Then due to a admin priority change and the specific `t_submit` of that job, a job needs to go between the job with queue priority 100 & 99.  Don't all of the other jobs have to have their priorities adjusted as well?

**Comment by grondo** [comment]
> Don't all of the other jobs have to have their priorities adjusted as well?

I think you answered your own question. In the simple case the priority value is a function of `t_submit` and admin priority, not a function of other jobs in the queue. Simply, you calculate a new priority for a job and then insert that job into the queue in the order dictated by that priority.

> a bunch of jobs are assigned queue priorities of 100, 99, 98, 97, 96. Then due to a admin priority change and the specific t_submit of that job, a job needs to go between the job with queue priority 100 & 99. Don't all of the other jobs have to have their priorities adjusted as well?

I don't think the priority values of adjacent jobs would ever be sequential, and priority "ties" would be almost astronomically unlikely (`t_submit` as a factor  in the priority calculation should eliminate that possibility). That is why the queue priority is given a huge range (or alternately a floating-point value between 0. and 1.0, though we've chosen not to do that here)

**Comment by chu11** [comment]
> I don't think the priority values of adjacent jobs would ever be sequential, and priority "ties" would be almost astronomically unlikely (t_submit as a factor in the priority calculation should eliminate that possibility). That is why the queue priority is given a huge range (or alternately a floating-point value between 0. and 1.0, though we've chosen not to do that here)

Ok, perhaps I'm thinking more about the corner cases rather than the average case.

**Comment by garlick** [comment]
Agree with @grondo here.

I don't think we want to assert that priorities _must_ be unique though.  Maybe practically speaking we do need to be prepared to break ties deterministically, like order on priority, then job ID.  That would be a good subtlety to point out in an RFC at some point, if it makes sense.

**Comment by grondo** [comment]
It might have been less confusing if we kept priority as a floating point number between 0. and 1.0?

**Comment by grondo** [comment]
> like order on priority, then job ID.

Good point, I didn't mean that priority has to be unique, just that it is a function of factors that are consider on a per-job basis, not based on other jobs (except where they may influence those factors out-of-band, e.g. total number of queued jobs for a user). And the resulting priority value should have a large enough range to avoid ties becoming _common_.

BTW, by sorting ties on jobid second, there is a not-so-hidden benefit to submitting your job on the lowest rank possible ;-)

**Comment by dongahn** [comment]
> a simple callback into the plugin which given a job returns a priority. This would be used in place of any default job_update_priority() or similar function in the job manager. In this case the function would be called once per job, whenever the job manager wanted to update a job's priority. A limitation would be that the function to calculate the priority is understood not to block.

> a setpriority() call that the plugin could call asynchronously on any job. The job manager would then perhaps notify the priority plugin of job/jobs that need a priority update, and the plugin could either synchronously or asynchronously update the job or job's prioirty.

One comment: I think it will be useful to design the plugin API so that the plugin itself doesn't have to keep its own pending job queues. The duplication of states like this can lead to higher complexity,

**Comment by dongahn** [comment]
> It might have been less confusing if we kept priority as a floating point number between 0. and 1.0?

I fear normalizing the priority to 0. - 1.0 can lead to an issue down the load when we optimize to minimize the number of updates (e.g., reusing the old fair share values as much as possible: https://github.com/flux-framework/flux-accounting/issues/69). 

A question: do you see the plugin API will pass the primary priority down to the plugin so that it can incorporate that into the resulting priority? Or that is the responsibility of the job manager and the plugin should just concern the secondary priority. It seems the latter is more straightforward.

**Comment by grondo** [comment]
> One comment: I think it will be useful to design the plugin API so that the plugin itself doesn't have to keep its own pending job queues. The duplication of states like this can lead to higher complexity,

Yeah, I had assumed the plugin would not keep its own queue, and would instead keep the minimum state necessary for whatever factors it may be using to compute the priority value of any job. In many cases, the plugin may not need to keep any state if it is only using factors that are part of a job.

**Comment by garlick** [comment]
> A question: do you see the plugin API will pass the primary priority down to the plugin so that it can incorporate that into the resulting priority? Or that is the responsibility of the job manager and the plugin should just concern the secondary priority. It seems the latter is more straightforward.

We're getting away from primary and secondary priority as terms, but yes I think the administrative priority and the previous value of the queue priority would be available to the plugin via some kind of get function.

**Comment by dongahn** [comment]
> BTW, by sorting ties on jobid second, there is a not-so-hidden benefit to submitting your job on the lowest rank possible ;-)

I would argue we use submit time as our tie breaker since at least we can explain this to users more easily and I have to think this will have least bias. However, since our FLUID is loosely ordered by submit time, it may work just as well. But in case we need to explain this to users, we should say "submit time" since it is earlier to grok.

**Comment by grondo** [comment]
> I fear normalizing the priority to 0. - 1.0 can lead to an issue down the load when we optimize to minimize the number of updates (e.g., reusing the old fair share values as much as possible: flux-framework/flux-accounting#69).

Yeah, sorry I was mainly proposing that as a thought experiment as I think it would show the unlikelihood of the corner case example. However, we should point out the distinction between fair share values that the plugin might be using as _one_ of the factors in the calculation of priority. How the fair share table gets updated in the plugin is a different issue, right?

**Comment by dongahn** [comment]
> We're getting away from primary and secondary priority as terms, but yes I think the administrative priority and the previous value of the queue priority would be available to the plugin via some kind of get function.

I think my main question is whether the plugin should be responsible for incorporating the administrative priority of the job into the resulting priority. If this should be the case, we need a pretty disciplined way in the resulting priority. (e.g., using the high order bits of the result for the admin priority and low order bits for secondary priority).

**Comment by grondo** [comment]
> I think my main question is whether the plugin should be responsible for incorporating the administrative priority of the job into the resulting priority.

I think yes! This is one of the important factors of a multi-factor priority right?

> If this should be the case, we need a pretty disciplined way in the resulting priority. (e.g., using the high order bits of the result for the admin priority and low order bits for secondary priority).

This part confused me though, what do you mean by secondary priority here?

Would it help if we renamed the "admin priority" something like the "nice" value? (thought that would reverse its meaning)

Then there is just one priority, which is always calculated.

**Comment by garlick** [comment]
> I would argue we use submit time as our tie breaker since at least we can explain this to users more easily

I suggested FLUIDs because they are guaranteed unique.  It's possible to still have a tie with t_submit.

**Comment by dongahn** [comment]
> It's possible to still have a tie with t_submit.

Huh. I didn't know. I assumed each job will get to a central location at which point we time stamp. Clearly this assumption was incorrect... (t_submit is time stamped by the local broker in a distributed fashion, or similar?)

**Comment by garlick** [comment]
> t_submit is time stamped by the local broker in a distributed fashion, or similar

It's set from `gettimeofday()` (wall clock) in the ingest module approximately when it is received.  The ingest module runs across all ranks.

**Comment by dongahn** [comment]
>  How the fair share table gets updated in the plugin is a different issue, right?

Yes this is a different issue.

**Comment by dongahn** [comment]
> This part confused me though, what do you mean by secondary priority here?

> Would it help if we renamed the "admin priority" something like the "nice" value? (thought that would reverse its meaning)

> Then there is just one priority, which is always calculated.

We discussed this a bit at our coffee hour. I believe we converged that we don't want to mandate that the priority resulting from the plugin should be capable of strict lexicographical ordering of multiple factors although it can.

**Comment by garlick** [comment]
> Would it help if we renamed the "admin priority" something like the "nice" value? (thought that would reverse its meaning)
> 
> Then there is just one priority, which is always calculated.

Not a bad idea, though a bit of extra work.  We should probably decide thumbs up or down on this idea before merging #3339 (which is about ready to go if we don't make this change).  Probably we should go with a new range of -20 (un-nice) to +20 (nice) like the system one if we rename it to avoid confusion.

What do people think?  I don't really have a problem with the current "admin priority" and "queue priority" names, personally.

**Comment by grondo** [comment]
> What do people think? I don't really have a problem with the current "admin priority" and "queue priority" names, personally.

I don't really have a problem with it either, though I do worry about future confusion with multiple things called priority.

I do prefer "priority" as a concept to "nice" value since it is numerically more intuitive. (though our "default" admin priority of 16 is kind of unintuitive to me)

**Comment by dongahn** [comment]
I don't have a problem as far as everyone can look at the same page about what it is. The thing that assured me during our coffee hour is that, there are other ways for an admin to override the priorities calculated from  the plugin to make a job ahead of anything else.

**Comment by chu11** [comment]
While looking at https://github.com/flux-framework/flux-core/issues/3327, I realized a potential issue.

When the `job-info` module is initialized, it reads from the job eventlog to recreate job state.  But RFC21 says we shouldn't put the `priority` event in the eventlog.  So there's no way to know the "queue priority" of a job from the eventlog.

No biggie if we're reloading `job-info` and `job-manager` at about the same time (i.e. bootstrapping a instance), we normally get that from the journal coming from the job-manager.

But if we're just reloading the `job-info` module, we may never get a `priority` event in the journal.  Do we need some way to ping the `job-manager` and say "hey, send out this info in a journal event"?  Without the queue priority, `job-info` may not be able to sort jobs properly.  Which maybe isn't the biggest deal?

**Comment by garlick** [comment]
Yeah as mentioned in #3327, I agree with your assessment and think we should have two separate events.
I was thinking that we wouldn't need the event in the eventlog because `job-manager` should just flip back to PRIORITY from SCHED  during a replay, and let the priority plugin recalculate the priority.  But a state-changing event must be in the persistent eventlog or our design is broken!   I'll submit a change to the RFC.  Sorry for the misstep.

**Comment by chu11** [comment]
idea - for the alloc event, perhaps we log the "final" queue priority for the job before it was run?  so that when on restart `job-info` will be able to load that data?

**Comment by garlick** [comment]
I call a design issue pause (on myself)!

If we post the `priority` event to the eventlog, then when the job manager restarts, the eventlog replay will cause it to skip the PRIORITY state and not recalculate the initial priority.  So the queue order would be wrong until the first reprioritization pass.

If we've said the priority plugin can do some initialization that might involve phoning home to flux-accounting the first time it sees a user (and that was one motivation for the PRIORITY state) but NOT after that, how does that work after a restart if the initial priority came from the eventlog?

On a restart, if job-info replays the eventlog then applies the journal, it won't get the most up to date priority unless the last round of queue prioirty updates fits in the journal.

**Comment by garlick** [comment]
I think one misstep was allowing the SCHED state to regress to PRIORITY state across a job manager restart without a transitioning event.  If we added, say, a `priority-invalidate` event that the job manager posts when it replays the eventlog of a job and finds it in the SCHED state, then a) job manager would contact the priority plugin again to re-establish the priority, and b) external entities like job-info can accurately track the job state.

RFC 21 has these rules
* Replaying the job eventlog SHALL accurately reproduce the current job state.
* The job state machine SHALL NOT contain cycles, to avoid ambiguity in synchronization.

We are violating both of these rules now.  We would not be violating the first one if add `priority-invalidate`.
We could delete the no-cycles rule or soften it to "...SHOULD NOT contain cycles..."

On synchronizing job-info, first I don't think the queue priority should be considered valid when the job is not in SCHED state (in a sense it is not in a queue at that point).  I do think, for testing if nothing else, users should be able to fetch the priority value, for example via a custom `flux jobs` format, but it would be appropriate to print `-` or similar if the job is not in SCHED state.

Does that make sense?

**Comment by grondo** [comment]
> We could delete the no-cycles rule or soften it to "...SHOULD NOT contain cycles..."

Maybe step back and think about why you wanted to avoid cycles originally. If it seems important, then I would keep the rule and figure out how to avoid the cycle. If not, then just delete the no-cycles rule. Stating SHOULD NOT then doing it anyway in the same RFC could be considered a bit odd.

**Comment by garlick** [comment]
> Maybe step back and think about why you wanted to avoid cycles originally. If it seems important, then I would keep the rule and figure out how to avoid the cycle. If not, then just delete the no-cycles rule. Stating SHOULD NOT then doing it anyway in the same RFC could be considered a bit odd.

Yeah that's true, I probably shouldn't have suggested that option.

I do remember one issue we discussed was allowing a cycle from RUN to SCHED state for _grow_.  I was against this as it seemed to me it would make it much harder to reason about and synchronize on things like whether or not the guest namespace existed, or whether a job had resources or shells.  I'm not sure the blanket statement was appropriate though.   I would lean towards deleting it, and deciding the impact of each proposal to add a cycle on a case by case basis.

I may have overstated the case in order to make the design more understandable at the time (back when there were infinite possibilities!)

I hope I'm not arguing for this now just because I'm not being creative.

**Comment by grondo** [comment]
That sounds reasonable to me.

It would impossible to avoid cycles if any job state needs to be re-obtained after a job-manager restart.
That makes me wonder, would a more generic `reload` event or similar be more useful than `priority-invalidate`, or do you think priority is the only job state that would need to be invalidated on reload?

I wonder if in the future we'll want to be able to revert back to DEPEND state for jobs for some reason, e.g. if we allow job updates and a new dependency or minimum job begin-time is applied.

Sorry, I'm probably not helping here.

**Comment by garlick** [comment]
Oh I view the eventlog replay as a process that restores the current state, not actually re-acquiring those states. For example, during the replay of an inactive job, we don't send an `alloc` request to the scheduler as it passes through SCHED state.

We actually want the job to transition from SCHED back to PRIORITY on a restart because we want the priority plugin to be contacted in that conext to allow for an asynchronous response.  Now maybe that design point could be separately debated.  I think the result of the previous discusson on it was that we should allow for the plugin to go get info about a user that it doesn't know about (or at least the possibility of that) before the job becomes schedulable.

Does that make sense and did I understand your point?

> I  wonder if in the future we'll want to be able to revert back to DEPEND state for jobs for some reason, e.g. if we allow job updates and a new dependency or minimum job begin-time is applied.

Yeah a transition from ALLOC back to DEPEND probably does make sense in that case IMHO.  So I guess that's a +1 in favor of dropping the no-cycles rule.

**Comment by chu11** [comment]
> On synchronizing job-info, first I don't think the queue priority should be considered valid when the job is not in SCHED state (in a sense it is not in a queue at that point). I do think, for testing if nothing else, users should be able to fetch the priority value, for example via a custom flux jobs format, but it would be appropriate to print - or similar if the job is not in SCHED state.

That sounds good.  Its definitely not needed except for testing / curiosity.  Perhaps queue priority is available on a "best effort" level for non-pending jobs.

> Edit: oh we still have the problem of getting the latest priority if job-info is reloaded, since the journal might have wrapped. Hmm.

Just throwing out an idea.  Could the `job-manager` be queried with a set of jobids and the queue priorities could be returned?  Or perhaps it simply informs the `job-manager` to write those queue priorities to the journal and doesn't even have to respond to the RPC?

**Comment by grondo** [comment]
Sorry, I misspoke, by "state" I meant job data acquired by some part of the
system which isn't available in the eventlog, the priority plugin
initialization being one example.

**Comment by garlick** [comment]
> by "state" I meant job data acquired by some part of the
system which isn't available in the eventlog, the priority plugin
initialization being one example.

You mean like if we had another state like PRIORITY used to acquire other data with the same semantics?  I guess then you could use the same `reload` event to transition out of either state.  That seems unlikely to me but I wouldn't be opposed to renaming `priority-invalidate` to `reload` or similar just in case.  I guess it would mean that we would want to post a `reload` event to every active job, not just those in SCHED state.  Hmm, actually, that could well be a useful thing to log for running jobs too, for diagnosing future issues.  OK, I kind of like that!

**Comment by garlick** [comment]
> Just throwing out an idea. Could the job-manager be queried with a set of jobids and the queue priorities could be returned? Or perhaps it simply informs the job-manager to write those queue priorities to the journal and doesn't even have to respond to the RPC?

Yeah I was mulling around ideas like that, but wanted to be sure we don't a) pollute the journal concept, or b) create more ordering issues that increase complexity for journal consumer.

**Comment by chu11** [comment]
> Yeah I was mulling around ideas like that, but wanted to be sure we don't a) pollute the journal concept, or b) create more ordering issues that increase complexity for journal consumer.

Slightly alternate idea, RPC to the job-manager tells it to replay the eventlog journal for a specific jobid.

**Comment by garlick** [comment]
> Slightly alternate idea, RPC to the job-manager tells it to replay the entire eventlog journal for a specific jobid.

Except we don't currently keep any of the events around in the job manager...

I was just wondering if we should get rid of the JOURNAL_ONLY event idea and instead have a streaming RPC that tracks requested non-persistent attributes like priority?

You'd know better than I what complications would arise when trying to consume data from both, when they are only loosely synchronized.  I guess you'd have to deal with a priority update before you know about the job from the journal.  Maybe if those things were kept in a separate hash in job-info that's accessed on demand (rather than having an RPC in the job-manager to provide them on demand?)  Could we go one further and put that "attribute hash" or priority hash or whatever in a reusable libjob container?

I think we've gone off topic for the plugin interface, and this issue is getting quite long.    I'll open a new issue.

**Comment by garlick** [comment]
Summary of points discussed by @grondo and @garlick offline:
* let's focus on minimal priority plugin interface now, while planning for generalized job manager plugins in the future
* there should be a default priority "builtin plugin" 
* TOML config can select loaded job manager plugins:
  -  load a glob of plugin paths: `[job-manager] plugins = "glob"`
  -  per plugin config also: `[job-amanger.plugin.priority] accounting-db = "uri"`
*  let's reuse `<flux/plugin.h>` to define plugin interfaces
*  callback:  new job received (plugin may set priority asynchronously)
* callback: update priority (plugin must respond immediately, e.g. this s a function call)
* plugin may request job manager to update all jobs' priorities;  job manager would then iterate over priority update
* a job manager handle with accessor to module `flux_t`  will be needed
* a handle representing a job will be needed
* plugin may store arbitrary data with job manager or job handles

I might have missed a few nuances!  Please fill in @grondo

**Comment by grondo** [comment]
Looks pretty good to me!

>  * callback: new job received (plugin may set priority asynchronously)
>  * callback: update priority (plugin must respond immediately, e.g. this s a function call)

Will a priority plugin also need a callback when a job is inactive? e.g. if it has some internal state related to jobs it will need to know when it can free that state (assuming the state is not directly associated with the job handle, in which case the state will be auto-destroyed).

Therefore, for future extensibility, I'd propose a set of `job.` callbacks, starting with `job.new` and `job.inactive` here. Maybe it would make more sense to call these job event callbacks, which are issued when the job transitions to the named event. Later, a callback for each job state could be added, along with other events like `job.update` (but that is for the generic `job-manager` plugin case)

**Comment by dongahn** [comment]
Here are some of my random notes from today's coffee hour so that @cmoussa1 will have some context for tomorrow:

- Each job state transition even will be passed to the priority plugin. Expectation is that the plugin will use the information to extract requisite global metadata data (e.g., how many running jobs a user has and use that to enforce running job limit etc)

- The jobspec for each job will be passed to the plugin as well. If the user uses non-default bank, the plugin can find the bank info the jobspec's `attributes` section. Different banks will likely produce different priorities. (TODO: Ask Ryan Day to see  how important it is to support a single use multiple banks configurations at LLNL.)

- There needs to be a way for a priority plugin to be able to register an RPC -- this way, the plugin can use it to receive a bulk fairshare update from flux-accounting on each priority recalculation. RPC needs to be highly performing utilizing asynchronous computations as much as possible. You don't want this third-parity RPC to head-of-line-block other critical  RPCs of job-manager. We will need a benchmark to measure its performance (scaling the job count to 10K or similar). 

- A job needs to be able to get its priority updated while it is in either PRIORITY or SCHED state.

**Comment by chu11** [comment]
Began working on a prototype implementation for #3325, mostly to get my head around `libflux/plugin.h` b/c I didn't know the API too well.   Some design thoughts/questions I hit along the way (dunno if this should be spun off into a different issue or discussion, posting here for now).

> callback: update priority (plugin must respond immediately, e.g. this is a function call)

I was imagining a plugin callback that was something like:
                                                                                                                                            
```                                                                                                                                         
int update_priority (int urgency, uint64_t t_submit, unsigned int *priority);
```                                                                                                                                         

or with our job-manager job handle

```
int update_priority (int urgency, uint64_t t_submit, job_manager_job_t *job);
```
                                                                                                                                            
But with the plugin handler call:
                                                                                                                                            
```                                                                                                                                         
int flux_plugin_call (flux_plugin_t *p, const char *name,                                                                                   
                      flux_plugin_arg_t *args);                                                                                             
```                                                                                                                                         

I don't think you can pass a pointer.  There are non-optimal ways to do it:

- via the `void *arg` you set when the handler is created you could set the "job-manager handle", the job could be looked up in the job-manager context and set there.  But having the user lookup the job seems like more work than we should expect them to do?  Especially for the default priority plugin, which is just going to return the same urgency it was given.
- via an aux set/get, which isn't the most obvious/expected way to do it
- hackish way, pass pointer via `{s:I}` in the args, I don't think that's a good idea

Almost feel like `flux_plugin_f` should be:

```
typedef int (*flux_plugin_f) (flux_plugin_t *p,
                              const char *topic,
                              flux_plugin_arg_t *args,
                              void *handler_data,
                              void *call_data);
```

`handler_data` being the "global" arg set at handler creation time and `call_data` set on each call.

> callback: new job received (plugin may set priority asynchronously)

Using the current plugin API and the `flux_plugin_call()`, wasn't obvious how to do this.  I assumed a prep-check reactor in the plugin to occasionally process new jobs and then call some `job_manager_set_job_priority (flux_jobid_t id, unsigned int priority)`?

Perhaps would be easier to setup a service that the job-manager would send RPCs too?

Both not the easiest API for a plugin writer.  But we can add a convenience library similar to `libschedutil` that would hide gory details.

**Comment by garlick** [comment]
Uh oh, I think @grondo already has a prototype working for this.  I'll let him respond to your comments on the plugin.h interface.  I'm pretty sure we want the job manager to do all lookups/iterations on the job hash on behalf of the plugin (or unbeknownst to the plugin).

**Comment by grondo** [comment]
Yeah, sorry @chu11, before the break I spent a little time on a job manager plugin API which would hopefully be sufficient to develop a priority plugin. The intent was to just come up with the plugin-side of the API (e.g. equivalent to `flux/shell.h`), but that ended up being difficult to do without also making a prototype - though the prototype ended up being quite simple.

First some general comments on the `plugin.h` API:

> Almost feel like flux_plugin_f should be:
>
> ```
> typedef int (*flux_plugin_f) (flux_plugin_t *p,
>                               const char *topic,
>                               flux_plugin_arg_t *args,
>                               void *handler_data,
>                               void *call_data);
> ```
> handler_data being the "global" arg set at handler creation time and call_data set on each call.

`void *` args to plugin callbacks were purposely left out of the API. When raw pointers to internal data are passed as plugin args, it becomes too tempting to pass pointers to structs which might change between different versions of Flux, thus meaning that a plugin will only work with the version of flux-core it was compiled against. We want to avoid that when possible.

There's a way around this though, and that is to share opaque "handles" with the plugin via the `flux_plugin_t *p` aux container. This is what is done in the shell plugin API:

```c
/*  Get flux_shell_t object from flux plugin handle
 */
flux_shell_t * flux_plugin_get_shell (flux_plugin_t *p);

/*  Return the current task for task_init, task_exec, and task_exit callbacks:
 *
 *  Returns NULL in any other context.
 */
flux_shell_task_t * flux_shell_current_task (flux_shell_t *shell);
```
This is safer and more efficient since you aren't passing the `flux_shell_t *` and `flux_shell_task_t *` as `void *arg`, and they are only fetched as needed.

> > callback: new job received (plugin may set priority asynchronously)
>
> Using the current plugin API and the flux_plugin_call(), wasn't obvious how to do this. I assumed a prep-check reactor in the plugin to occasionally process new jobs and then call some job_manager_set_job_priority (flux_jobid_t id, unsigned int priority)?

In the prototype, I simply added a call to the plugin to each new state transition, before `event_job_action()`. A priority plugin could then subscribe for a `state.priority` callback, which would indicate a job is entering the PRIORITY state. The plugin would then either set a priority immediately by pushing a `priority` on the `FLUX_PLUGIN_ARG_OUT`. If it doesn't set a priority immediately, the job would stay in PRIORITY state until the plugin asynchronously sets a new priority. How this is done exactly is TDB, but I think @garlick agreed with your idea that a new RPC for setting job priority would work (it would be job-manager sending an RPC to itself), or we could add a new function to the plugin API like

```C
int flux_plugin_set_job_priority (flux_plugin_t *p, flux_jobid_t id, unsigned int priority);
```

Which would trigger the job manager to move this job out of PRIORITY state on the next check/prep watcher or something.

The other open question is how to allow the plugin to request a reprioritization of all jobs. I think we had said the plugin would request reprioritization and then the job manager would call a `priority.set` or similar callback for each job in its own good time.

The current prototype is here (without any support for asynchronous priority setting in PRIORITY state, nor a way to request a reprioritization of all jobs)

https://github.com/grondo/flux-core/commit/b57858fab159e86759778c979af15d123870e946

The prototype calls job-manager plugins "jobtap" plugins (because flux_job_manager_plugin.h was too long, and also because they "tap" into job state transitions)

Instead of adding a `flux_jobtap_job_t *` handle with getters for each bit of job information, the jobtap interface passes all the job data as args in the `flux_plugin_arg_t`, e.g.:

```C
    flux_plugin_arg_t *args = flux_plugin_arg_create ();
    flux_plugin_arg_pack (args,
                          FLUX_PLUGIN_ARG_IN,
                          "{s:O s:I s:i s:i s:i s:I s:f}",
                          "jobspec", job->jobspec_redacted,
                          "id", job->id,
                          "userid", job->userid,
                          "urgency", job->urgency,
                          "state", job->state,
                          "priority", job->priority,
                          "t_submit", job->t_submit);
```

The prototype includes 3 "builtin" plugins, e.g. the following is the `builtin.default` plugin:

```c
/*  The current implementation of priority.set just copies
 *   the urgency to the priority.
 */
static int priority_cb (flux_plugin_t *p,
                        const char *topic,
                        flux_plugin_arg_t *args,
                        void *data)
{
    int urgency = 16;
    flux_plugin_arg_unpack (args, FLUX_PLUGIN_ARG_IN,
                            "{s:i}",
                            "urgency", &urgency);
    flux_plugin_arg_pack (args, FLUX_PLUGIN_ARG_OUT,
                          "{s:i}",
                          "priority", urgency);
    return 0;
}

int default_priority_plugin_init (flux_plugin_t *p)
{
    if (flux_plugin_set_name (p, "priority") < 0
        || flux_plugin_add_handler (p,
                                    "state.priority",
                                    priority_cb,
                                    NULL) < 0
        || flux_plugin_add_handler (p,
                                    "priority.set",
                                    priority_cb,
                                    NULL) < 0) {
        return -1;
    }
    return 0;
}

```

and  a `hold` plugin, which sets prioirty=0 on all incoming jobs:

```c
/* Always set priority=0 so all jobs are submitted in held state
 */
static int hold_cb (flux_plugin_t *p,
                        const char *topic,
                        flux_plugin_arg_t *args,
                        void *arg)
{
    flux_plugin_arg_pack (args, FLUX_PLUGIN_ARG_OUT, "{s:i}", "priority", 0);
    return 0;
}


int hold_priority_plugin_init (flux_plugin_t *p)
{
    return flux_plugin_add_handler (p, "state.priority", hold_cb, NULL);
}
```

Finally, here's a plugin that subscribes to all jobtap callbacks and dumps all available info for each job:
```c
static int cb (flux_plugin_t *p,
               const char *topic,
               flux_plugin_arg_t *args,
               void *arg)
{
    json_t *resources;
    flux_jobid_t id;
    uint32_t userid;
    int urgency;
    unsigned int priority;
    flux_job_state_t state;
    double t_submit;
    char *s;

    flux_plugin_arg_unpack (args, FLUX_PLUGIN_ARG_IN,
                            "{s:{s:o} s:I s:i s:i s:i s:i s:f}",
                            "jobspec", "resources", &resources,
                            "id", &id,
                            "userid", &userid,
                            "urgency", &urgency,
                            "priority", &priority,
                            "state", &state,
                            "t_submit", &t_submit);

    s = json_dumps (resources, JSON_COMPACT);
    fprintf (stderr, "%s: id=%ju uid=%u urg=%d pri=%u state=%d t_submit=%.5f\n",
             topic, id, userid, urgency, priority, state, t_submit);
    fprintf (stderr, "resources: %s\n", s);
    free (s);
    return 0;
}

int flux_plugin_init (flux_plugin_t *p)
{
    return flux_plugin_add_handler (p, "*", cb, NULL);
}

```


For testing purposes, the branch allows plugins to be loaded via a new `job-manager.jobtap` RPC. The builtin "demo" plugins can be loaded as `builtin.default`, `builtin.age`, or `builtin.hold`, or a path to a DSO can be specified. Use the following python script:

```python
import flux
import sys
print(flux.Flux().rpc("job-manager.jobtap", {"load": sys.argv[1]}).get())
```

The special `none` argument will unload the current jobtap plugin and disable jobtap (again for testing purposes).

**Comment by chu11** [comment]
>  it becomes too tempting to pass pointers to structs which might change between different versions of Flux, thus meaning that a plugin will only work with the version of flux-core it was compiled against. We want to avoid that when possible.

Yup I agree.  I had read "plugin may store arbitrary data with job manager or job handles" in the design notes above as meaning having access to `struct job *` and such.  But it meaning access in an opaque getter/setter manner makes more sense.

> The plugin would then either set a priority immediately by pushing a priority on the FLUX_PLUGIN_ARG_OUT. 

Ahhh, this was the big subtlety I missed in the plugin API.  That the args parameter could be in & out.  That's the way a value could be passed back to the caller.

>  If it doesn't set a priority immediately, the job would stay in PRIORITY state until the plugin asynchronously sets a new priority. How this is done exactly is TDB, but I think @garlick agreed with your idea that a new RPC for setting job priority would work (it would be job-manager sending an RPC to itself), or we could add a new function to the plugin API like
>  
>
> int flux_plugin_set_job_priority (flux_plugin_t *p, flux_jobid_t id, unsigned int priority);
>
> Which would trigger the job manager to move this job out of PRIORITY state on the next check/prep watcher or something.


It seems you reached the same TBD point that I did.  I like the idea of the `flux_plugin_set_job_priority()` like function, but I kept on coming back to trying to make sure the plugin writer would never have to setup their own prep/check to do asynchronous priority calculations.  But perhaps it is inevitable that a plugin writer would have to on occasions.

**Comment by grondo** [comment]
>  but I kept on coming back to trying to make sure the plugin writer would never have to setup their own prep/check to do asynchronous priority calculations. But perhaps it is inevitable that a plugin writer would have to on occasions.

Yeah, IIUC  in _most_ cases a priority plugin would be able to return a priority immediately on entry to the PRIORITY state, so asynchronous exit from this state would be the uncommon case.

The use case I believe we were considering is that a fair share plugin might have to contact a database or other service when, for example, it sees the first job for a given user. In this case, the plugin might send an RPC to a remote service to fetch data for that user, and then would update the job priority in the continuation callback of that RPC, when it has the necessary information.

I don't see a need for the plugin itself to register prep/check or idle watchers in this case. The job manager itself should be responsible for any work that should be done in this manner (if necessary).

**Comment by chu11** [comment]
> The use case I believe we were considering is that a fair share plugin might have to contact a database or other service when, for example, it sees the first job for a given user. In this case, the plugin might send an RPC to a remote service to fetch data for that user, and then would update the job priority in the continuation callback of that RPC, when it has the necessary information.

I was also thinking of a case where the plugin wanted to "batch" calculate a bunch of priorities.  Perhaps the database needs to do some big calculation or something so we're going to wait awhile.  Perhaps this is not a common case and we don't have to think about it too much.

**Comment by garlick** [comment]
As I recall the idea we talked about was to have some way afor the plugin to notify the job manager that it wants to iterate over all jobs (or perhaps all jobs in a particular set of states), callling the plugin callback for each job.  That would let the job manager do the prep/check/idle, if necessary to avoid becoming unresponsive if, say, it needs to iterate over a million jobs.

**Comment by chu11** [comment]
> As I recall the idea we talked about was to have some way for the plugin to notify the job manager that it wants to iterate over all jobs (or perhaps all jobs in a particular set of states), calling the plugin callback for each job. That would let the job manager do the prep/check/idle, if necessary to avoid becoming unresponsive if, say, it needs to iterate over a million jobs.

I was thinking more about the case a bajillion new jobs were submitted in short order.  So the priority plugin got a bunch of `job.new` notifications in short order.

**Comment by garlick** [comment]
Oh I see.  Hmm, well I guess that sort of thing would be possible since the plugin has access to the flux_t handle and reactor, but the idea would be to minimize the need to place less burden on the plugin writer.

**Comment by chu11** [comment]
> Oh I see. Hmm, well I guess that sort of thing would be possible since the plugin has access to the flux_t handle and reactor, but the idea would be to minimize the need to place less burden on the plugin writer.

Its certainly not a common case and probably one we don't need to fret about too much right now.  I'm sure my thoughts were just wandering over many ideas over the long weekend :-)

**Comment by grondo** [comment]
> I was thinking more about the case a bajillion new jobs were submitted in short order. So the priority plugin got a bunch of job.new notifications in short order.

The batching could be done in a plugin by having internal state that keeps a list of jobs which need their priority updated if an asynchronous update is in progress.

For example, the first job `state.priority` callback triggers an RPC to a database service and puts the current job as the first job on a list of pending updates. While the RPC is in progress, new jobs are simply appended to the internal list. Once the RPC continuation callback runs, the priority plugin could iterate over all jobs recalculating their priorities and calling e.g. `flux_plugin_set_job_priority(p, jobid, priority)` for each job.

Otherwise, if no RPC is in progress and the priority can be calculated immediately (i.e. all factors are currently available), then the plugin would just calculate and immediately return the priority.

Note that a plugin could also schedule a periodic database query by using a timer or periodic watcher on the flux handle. That RPC could be treated the same as above, or once it completes, the plugin could request a priority recalc for all jobs.

**Comment by garlick** [comment]
Oh duh, that could be very common.  Good discussion!

**Comment by grondo** [comment]
Should have been closed by #3464, though there may be some unimplemented thoughts from this issue we want to push off to new issues. (Feel free to reopen if desired)
